### PR TITLE
profile manipulation + analysis scripts

### PR DESCRIPTION
**Pull request by kgrosman**
Includes: profile intersections/unions (and more nuanced test files), converting back and forth between profiles and their dumps, switching method tags, identifying tag combination counts among methods that do/don't have their corresponding in the profile, and verifying that a particular profile is valid and can be used by ART.

### CODE REVIEW

File: CUJExecutor/back [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+# INPUT: a .prof file followed by any number of tags in {HSPL, HPL, PL}
```
Comment (diablonhn) [inline CUJExecutor/back]: Technically, the tags are:

  - H
  - S
  - P

A method in the profile can be prefixed with any or all of them, in any order.  But it's okay (though not safe) to assume that the tags only come together as `HSPL`, `HPL`, or `PL`.

`L` is not a flag and it is the start of a Java class signature that ends with `;`.

File: CUJExecutor/back [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/back]: The file `back` needs a more descriptive name.

File: CUJExecutor/buildProfile [hunk 3]
```diff
@@ -68,7 +68,7 @@ EMPTY_PROFILE="utilityProfiles/empty.prof" #location of an empty profile
 STARTUP_PROFILE="utilityProfiles/startup.prof" #location of an startup profile
 CUR_PROFILE="/data/misc/profiles/cur/0/com.google.android.apps.maps/primary.prof" #location of the cloud profile on the device
 TEMP_PHONE_PROFILE="/sdcard/temp_profile.prof"
-STASHING_COMPILED_ACTIONS="/tmp/manipulate_compiled_actions.txt" #Contains the actions used to compile, in case callee wants it
+STASHING_MPT_INFO="/tmp/manipulate_info.txt" #Contains the actions used to compile, in case callee wants it
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Should just use the unshortened name `STASHING_MANIPULATE_INFO`.

File: CUJExecutor/buildProfile [hunk 4]
```diff
@@ -138,50 +139,58 @@ for ((i=$COMPILATION_FLAG_SEQ_STARTING_IDX; i<${#COMPILATION_FLAG_SEQ}; i++))
...
+	if [ "$COMPILATION_FLAG" != "c" ] && [ "$COMPILATION_FLAG" != "p" ] && [ "$COMPILATION_FLAG" != "f" ] && [ "$COMPILATION_FLAG" != "m" ] && [ "$COMPILATION_FLAG" != "h" ] 
 	then	
 		echo "Builder actions flag: $COMPILATION_FLAG is invalid"
 		exit 1
 	fi
+	
+	if [ "$COMPILATION_FLAG" = "h" ] 
+	then
+		./convertToH $CREATED_PROFILE > "/tmp/tmp.txt"
+		cat "/tmp/tmp.txt" > $CREATED_PROFILE
```
Comment (diablonhn) [inline CUJExecutor/buildProfile]: Can this use `cp` instead?

File: CUJExecutor/churn [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+# INPUT: a .prof file
+# OUTPUT: recompiles the app with the passed profile and returns the ref profile (which might differ from the passed profile)
+
+PROFILE=${1?"ERROR:NEED TO PROVIDFE A PROFILE"}
```
Comment (diablonhn) [inline CUJExecutor/churn]: Typo `PROVIDFE`.

File: CUJExecutor/churn [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/churn]: Name `churn` isn't too clear what the file is doing - needs a better name.

File: CUJExecutor/churn [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+# INPUT: a .prof file
+# OUTPUT: recompiles the app with the passed profile and returns the ref profile (which might differ from the passed profile)
+
+PROFILE=${1?"ERROR:NEED TO PROVIDFE A PROFILE"}
+
+
+CHURNED_PROFILE="/tmp/churn_churned_profile.txt"
+
+./compileWith $PROFILE
+adb pull /data/misc/profiles/ref/com.google.android.apps.maps/primary.prof $CHURNED_PROFILE 2>&1 >/dev/null
```
Comment (diablonhn) [inline CUJExecutor/churn]: Optional: support other apps besides `com.google.android.apps.maps`.

File: CUJExecutor/compileWith [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/compileWith]: `compileWithProfile` is probably a better name than `compileWith`.

File: CUJExecutor/convert [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+#Prints out a version of this profile where all dump entries with the tag $FROM are switched to having the tag $TO
```
Comment (diablonhn) [inline CUJExecutor/convert]: Change `dump` to `profman classes-and-methods dump`?

File: CUJExecutor/convertToH [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/convertToH]: Spell out `convertToH` to `convertToHot`?

File: CUJExecutor/fileToDump [hunk 11]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+# INPUT: a .prof or .txt file
+# OUTPUT: a method-class dump. 
+# If a .txt file is passed, it is enterpretted as the dump to be returned.
```
Comment (diablonhn) [inline CUJExecutor/fileToDump]: Typo `enterpretted`?  Instead of interpretted, the file is simply passed through.

File: CUJExecutor/keep [hunk 12]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/keep]: Needs a better descriptive name than just `keep`.

File: CUJExecutor/methodClassBacking [hunk 13]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+FILE=${1?"ERROR: need to provide a file"}
+
+# INPUT: a .prof file or a method-class dump of a .prof file
+# OUTPUT: stores two dump files, one with all of the dump entries that have their corresponding class in the original profile and the other which holds all other entries
+#         echos the number of entries with each tag in each file
+
+
+#If we are provided a profile, convert it to it's dump. If we were given a text file, interpret it as the dump
```
Comment (diablonhn) [inline CUJExecutor/methodClassBacking]: `it's` -> `its`

File: CUJExecutor/methodClassBacking [hunk 14]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
Comment (diablonhn) [inline CUJExecutor/methodClassBacking]: `methodClassBacking` -> `addMethodClassToProfile`?

File: CUJExecutor/stableProfile [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+# INPUT: a .prof file
+# OUTPUT: a message saying whether the profile is stable.
```
Comment (diablonhn) [inline CUJExecutor/stableProfile]: What is the definition of a stable profile?

**Review by diablonhn — CHANGES_REQUESTED** [review]
Pull request needs a better name and description.

File: CUJExecutor/back (hunk 1 above)
Comment (kgrosman) [reply CUJExecutor/back]: Hmmm from what I understand, you could have any of these tags, but when you convert from a profile to a dump, they are always in the order H -> S -> P -> L and I've never seen a ref profile have any combinations other than HSPL, HPL, or PL. Is this accurate to your knowledge?

File: CUJExecutor/stableProfile (hunk 15 above)
Comment (kgrosman) [reply CUJExecutor/stableProfile]: definition on line 5 :P

File: CUJExecutor/back (hunk 1 above)
Comment (diablonhn) [reply CUJExecutor/back]: It makes sense for ART to only save certain tags to the profile, hence the lack of other combinations.

**Review by diablonhn — CHANGES_REQUESTED** [review]
There are a lot of scripTestFiles.  Is that intended?

**Comment by kgrosman** [comment]
> There are a lot of scripTestFiles. Is that intended?

Yeah, the test files were modified to be a little more nuanced (since I made a change to the way prof intersections/unions worked)

**Review by diablonhn — APPROVED** [review]

**Pull request merged**